src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+  #
+  # finish up all procedure strings
+  #
+  for key, s in proc_map.pairs():
+    result &= s
+    result &= "\n" # add a blank line between each proc
+
+
+macro db*(connection, user, password, database: string, body: untyped): untyped =
+  ##[
+    DB models definition. Models are defined as regular Nim objects in regular ``type`` sections.
+
+    ``connection``, ``user``, ``password``, ``database`` are the same args accepted by a standard ``dbConn`` instance.
+
+    The macro generates ``withDb`` template that wraps all DB interations.
+  ]##
+
+  result = newStmtList()
+
+  var dbObjReprs: seq[ObjRepr]
+  var normObjectNamesRegistry: seq[string] = @[] # this is later injected into context
+  var newCollections: seq[string]
+
+  for node in body:
+    if node.kind == nnkTypeSection:
+      let typeSection = node.ensureIdFields()
+      result.add typeSection
+
+      for typeDef in typeSection:
+        # echo typeDef.treeRepr
+        dbObjReprs.add typeDef.toObjRepr()
+
+    else:
+      result.add node
+
+  updateObjectRegistry(dbObjReprs)
+  for o in dbObjReprs:
+    newCollections.add o.getCollectionName
+
+  # echo $dbObjReprs
+  # echo $genObjectAccess(dbObjReprs)
+  # echo $genObjectToBSON(dbObjReprs)
+  # echo $genBSONToObject(dbObjReprs)
+
+
+  # let objectAccess =  parseStmt(genObjectAccess(dbObjReprs))
+  # result.add(objectAccess)
+
+  let withDbNode = getAst genWithDb(connection, user, password, database, -1, newCollections)  # -1 = true
+  result.insert(0, withDbNode)
+
+  let bsonToObject = parseStmt(genBSONToObject(dbObjReprs))
+  result.add(bsonToObject)
+
+  let objectToBSON = parseStmt(genObjectToBSON(dbObjReprs))
+  result.add(objectToBSON)
+
+macro dbAddCollection*(obj: typed): untyped =
+  ##[
+    Add a DB models for an object that has already been defined.
+
+    The macro generates ``withDb`` template that wraps all DB interations.
+  ]##
+  result = newStmtList()
+
+  let objName = $obj
+  if normObjectNamesRegistry.contains(objName):
+    raise newException(KeyError, "Macro dbAddCollection cannot add the same object again ($1).".format(objName))
+  normObjectNamesRegistry.add objName
+
+  var dbObjReprs: seq[ObjRepr]
+  var newCollections: seq[string]
+
+  let nSymbol = symbol(obj)
+  let typeDef = getImpl(nSymbol)
+  # echo typeDef.treeRepr
+  dbObjReprs.add typeDef.toObjRepr()
+
+  for o in dbObjReprs:
+    newCollections.add o.getCollectionName
+
+  # echo $genObjectToBSON(dbObjReprs)
+  # echo $genBSONToObject(dbObjReprs)
+
+  let withDbNode = getAst genWithDb("", "", "", "", 0, newCollections)
+  result.insert(0, withDbNode)
+
+  let bsonToObject = parseStmt(genBSONToObject(dbObjReprs))
+  result.add(bsonToObject)
+
+  let objectToBSON = parseStmt(genObjectToBSON(dbObjReprs))
+  result.add(objectToBSON)
+
+macro dbAddObject*(obj: typed): untyped =
+  ##[
+    Add an object for use a subtending JSON object. This macro does NOT create

```

Comment:
Sorry, I can't quite understand what you're saying here. Maybe, add an example?

Reply:
Adding more documentation now.